File: src/fortran/clipping.f90
module sutherland_hodgman
    use iso_c_binding
    implicit none
    
    type point
        real(kind=8) x
        real(kind=8) y
    end type point
    
    type vector
        real(kind=8) x
        real(kind=8) y
    end type vector
    
    type intersection_result
        logical success
        type(point) p
    end type intersection_result

contains

    function cross(U, V) result(cp)
        type(vector), intent(in) :: U, V
        real(kind=8) :: cp
        cp = U%x * V%y - U%y * V%x
    end function cross
    

    function dot(U, V) result(dp)
        type(vector), intent(in) :: U, V
        real(kind=8) :: dp
        dp = U%x * V%x + U%y * V%y
    end function dot
   

    subroutine push(array, n, value)
        integer, intent(inout) :: n
        real(kind=8), dimension(:, :), intent(inout) :: array
        type(point), intent(in) :: value
        array(1, n + 1) = value%x
        array(2, n + 1) = value%y
        n = n + 1
    end subroutine push
    

    subroutine copyto(src, dst, n)
        real(kind=8), dimension(:, :), intent(in) :: src
        real(kind=8), dimension(:, :), intent(inout) :: dst
        integer, intent(in) :: n
        integer :: i
        do i = 1, n
            dst(1, i) = src(1, i)
            dst(2, i) = src(2, i)
        end do
    end subroutine copyto
    

    function inside(p, r, U) result(is_inside)
        type(point) :: p, r
        type(vector) :: U
        logical :: is_inside
        is_inside = U%x * (p%y - r%y) > U%y * (p%x - r%x)
    end function inside
    

    function intersection(a, V, r, N) result(int_res)
        type(intersection_result) :: int_res
        type(point) :: a, r, p
        type(vector) :: V, N, W
        real(kind=8) :: nw, nv, t
        W = vector(r%x - a%x, r%y - a%y)
        nw = dot(N, W)
        nv = dot(N, V)
        if (nv /= 0.0d0) then
            t = nw / nv
            p = point(a%x + t * V%x, a%y + t * V%y)
            int_res = intersection_result(.true., p)
        else
            p = point(0.0d0, 0.0d0)
            int_res = intersection_result(.false., p)
        end if   
    end function intersection
    

    function polygon_area(polygon, length) result(area)
        real(kind=8) :: area
        type(point) :: a, b, c
        type(vector) :: U, V
        real(kind=8), dimension(:, :) :: polygon
        integer :: length, i
        
        area = 0.0d0
        a = point(polygon(1, 1), polygon(2, 1))
        b = point(polygon(1, 2), polygon(2, 2))
        U = vector(b%x - a%x, b%y - a%y)
        do i = 3, length
            c = point(polygon(1, i), polygon(2, i))
            V = vector(a%x - c%x, a%y - c%y)
            area = area + abs(cross(U, V))
            b = c
            U = V
        end do
        area = 0.5d0 * area
    end function polygon_area
    

    function clip_polygons(polygon, clipper, n_max) result(area)
        type(point) :: r, s, a, b, p
        type(vector) :: U, N, V
        type(intersection_result) :: int_res
        integer, dimension(2) :: polygon_shape, clipper_shape
        integer :: n_output, n_poly, n_clip, length, i, j
        integer(kind=8) :: n_max
        real(kind=8), dimension(:, :) :: polygon, clipper
        real(kind=8), dimension(2, n_max):: output, subject
        real(kind=8) :: area
        logical :: a_inside, b_inside
        
        polygon_shape = shape(clipper)
        clipper_shape = shape(polygon)
        n_poly = polygon_shape(2)
        n_clip = clipper_shape(2)

        n_output = n_poly
        call copyto(polygon, output, n_output)

        r = point(clipper(1, n_clip), clipper(2, n_clip))
        do i = 1, n_clip
            
            s = point(clipper(1, i), clipper(2, i))
            U = vector(s%x - r%x, s%y - r%y)
            N = vector(-U%y, U%x)
            
            if (U%x == 0.0d0 .and. U%y == 0.0d0) then
                cycle
            end if
            
            length = n_output
            call copyto(output, subject, length)
            n_output = 0
            
            a = point(subject(1, length), subject(2, length))
            a_inside = inside(a, r, U)
            do j = 1, length
                b = point(subject(1, j), subject(2, j))
                V = vector(b%x - a%x, b%y - a%y)
                
                if (V%x == 0.0d0 .and. V%y == 0.0d0) then
                    cycle
                end if
                
                b_inside = inside(b, r, U)
                
                if (b_inside) then
                    if (.not. a_inside) then
                        int_res = intersection(a, V, r, N)
                        if (int_res%success) then
                            call push(output, n_output, int_res%p)
                        end if
                    end if
                    call push(output, n_output, b)
                else if (a_inside) then
                    int_res = intersection(a, V, r, N)
                    if (int_res%success) then
                        call push(output, n_output, int_res%p)
                    else
                        b_inside = .true.
                        call push(output, n_output, b)
                    end if
                end if
                a = b
                a_inside = b_inside
            end do
            
            if (n_output < 3) then
                area = 0.0d0
                return
            end if
            
            r = s
        end do
        area = polygon_area(output, n_output)
    end function clip_polygons
    

    subroutine area_of_intersection(ndim, nvertex, ntriangles, polygons, clippers, areas) bind(c, name="area_of_intersection")
        !DEC$ ATTRIBUTES DLLEXPORT :: area_of_intersection
        integer(c_int64_t) :: ntriangles, nvertex, ndim, i
        real(c_double), dimension(ndim, nvertex, ntriangles), intent(in) :: polygons, clippers
        real(c_double), dimension(ntriangles), intent(out) :: areas
        do i = 1, ntriangles
            areas(i) = clip_polygons(polygons(:, :, i), clippers(:, :, i), nvertex * 2)
        end do
    end subroutine area_of_intersection

end

File: src/cplusplus/clipping.cpp
# include "clipping.h"


struct Point { double x, y; };
struct DVector { double x, y; };
struct IntersectionResult { bool succes; Point p; };


double cross_product(DVector U, DVector V) {
    return (U.x * V.y - U.y * V.x);
}


double dot_product(DVector U, DVector V) {
    return (U.x * V.x + U.y * V.y);
}


bool inside(Point p, Point r, DVector U) {
    return (U.x * (p.y - r.y) > U.y* (p.x - r.x));
}


IntersectionResult intersection(Point a, DVector V, Point r, DVector N) {
    DVector W = { r.x - a.x, r.y - a.y };
    double nw = dot_product(N, W);
    double nv = dot_product(N, V);
    if (nv != 0.0) {
        double t = nw / nv;
        Point p = { a.x + t * V.x, a.y + t * V.y };
        return { true, p };
    }
    else {
        Point p = { 0.0, 0.0 };
        return { false, p };
    }
}


double polygon_area(std::vector<Point> polygon, int length) {
    double area = 0.0;
    Point a = polygon[0];
    Point b = polygon[1];
    DVector U = { b.x - a.x, b.y - a.y };
    for (int i = 2; i < length; i++) {
        Point c = polygon[i];
        DVector V = { a.x - c.x, a.y - c.y };
        area += abs(cross_product(U, V));
        b = c;
        U = V;
    }
    return 0.5 * area;
}


void copyto(std::vector<Point> src, std::vector<Point> dst, int n) {
    for (int i = 0; i < n; i++) {
        dst[i] = src[i];
    };
};


double clip_polygons(std::vector<Point> polygon, std::vector<Point> clipper, int n_max) {
    std::vector<Point> subject(n_max), output(n_max);
    Point r, s, a, b;
    DVector U, V, N;
    bool a_inside, b_inside;
    int n_poly, n_clip, n_output, length;
    IntersectionResult int_res;

    n_poly = polygon.size();
    n_clip = clipper.size();
    n_output = n_poly;
    for (int ii = 0; ii < n_output; ii++) {
        output[ii] = polygon[ii];
    };

    r = clipper[n_clip - 1];
    for (int i = 0; i < n_clip; i++) {
        s = clipper[i];
        U = { s.x - r.x, s.y - r.y };
        N = { -U.y, U.x };

        if (U.x == 0.0 && U.y == 0.0) {
            continue;
        }

        length = n_output;
        for (int ii = 0; ii < length; ii++) {
            subject[ii] = output[ii];
        };
        n_output = 0;

        a = subject[length - 1];
        a_inside = inside(a, r, U);
        for (int j = 0; j < length; j++) {
            b = subject[j];
            V = { b.x - a.x, b.y - a.y };

            if (V.x == 0.0 && V.y == 0.0) {
                continue;
            }

            b_inside = inside(b, r, U);
            if (b_inside) {
                if (!a_inside) {
                    int_res = intersection(a, V, r, N);
                    if (int_res.succes) { 
                        output[n_output++] = int_res.p;
                    };
                }
                output[n_output++] = b;
            }
            else if (a_inside) {
                int_res = intersection(a, V, r, N);
                if (int_res.succes) {
                    output[n_output++] = int_res.p;
                } else {
                    b_inside = true; 
                    output[n_output++] = b; 
                }
            }
            a = b;
            a_inside = b_inside;
        }
        if (n_output < 3) return 0.0;
        r = s;
    }
    double area = polygon_area(output, n_output);
    return area;
}


void cpp_area_of_intersection(
    long ntriangles,
    long nvertex,
    long ndim,
    double* p,//olygons
    double* c,//lippers
    double* areas
) {
    long size = ndim * nvertex * ntriangles;
    std::vector<Point> polygon(nvertex), clipper(nvertex);
    for (auto i = 0; i < ntriangles; i++) {
        long start = i * 6;
        for (auto j = 0; j < nvertex; j++) {
            long first = start + j * 2;
            long second = start + j * 2 + 1;
            polygon[j] = { p[first], p[second] };
            clipper[j] = { c[first], c[second] };
        }
        areas[i] = clip_polygons(polygon, clipper, 2 * nvertex);
    }
}
